### ISSUE TITLE
compilation error 'nullptr'

### ISSUE THREAD

**Issue by vnitinv** [grpc/grpc#1051]
grpc-master]# uname -a
Linux gls-esx6-vm3 2.6.32-220.el6.i686 #1 SMP Tue Dec 6 16:15:40 GMT 2011 i686 i686 i386 GNU/Linux

make[1]: Leaving directory `/tmp/nitinkr/grpc-master/third_party/protobuf'
[CXX]     Compiling src/cpp/client/secure_credentials.cc
src/cpp/client/secure_credentials.cc: In function ‘std::unique_ptr<grpc::Credentials, std::default_delete<grpc::Credentials> > grpc::<unnamed>::WrapCredentials(grpc_credentials*)’:
src/cpp/client/secure_credentials.cc:69: error: ‘nullptr’ was not declared in this scope
src/cpp/client/secure_credentials.cc: In function ‘std::unique_ptr<grpc::Credentials, std::default_delete<grpc::Credentials> > grpc::SslCredentials(const grpc::SslCredentialsOptions&)’:
src/cpp/client/secure_credentials.cc:86: error: ‘nullptr’ was not declared in this scope
src/cpp/client/secure_credentials.cc: In function ‘std::unique_ptr<grpc::Credentials, std::default_delete<grpc::Credentials> > grpc::CompositeCredentials(const std::unique_ptr<grpc::Credentials, std::default_delete<grpc::Credentials> >&, const std::unique_ptr<grpc::Credentials, std::default_delete<grpc::Credentials> >&)’:
src/cpp/client/secure_credentials.cc:128: error: ‘nullptr’ was not declared in this scope
make: *** [objs/opt/src/cpp/client/secure_credentials.o] Error 1

**Comment by vnitinv** [comment]
just in case you ask be for -std being used. its using c++0x

/bin/sh ../libtool --tag=CXX   --mode=link g++ -pthread -Wall -Wwrite-strings -Woverloaded-virtual -Wno-sign-compare  -std=c++0x

**Comment by nicolasnoble** [comment]
What is your gcc version ?

**Comment by vnitinv** [comment]
gcc --version
gcc (GCC) 4.4.6 20110731 (Red Hat 4.4.6-3)

Note: I don't have an option to upgrade.

**Comment by nicolasnoble** [comment]
It's okay, it's most probably a #include missing. I'd be surprised if that compiler/stl don't have nullptr support in its C++0x section.

**Comment by vjpai** [comment]
I think the official #include is supposed to be #include <cstddef> , but it's probably included by default in most other headers. Maybe this particular one doesn't have it for this compiler

**Comment by vjpai** [comment]
Nitin, I can probably fix this but would you be willing to do the testing? I'm not going to be able to replicate your test environment, but I think I can make the code changes needed to work around the issues there. I'll attempt to solve this if you'll test it.

**Comment by ctiller** [comment]
https://gcc.gnu.org/projects/cxx0x.html

Looks like nullptr was added in 4.6. 4.4 was first released in 2009 and received patches thru 2012.

It's probably in the range of something we should support.

s/nullptr/NULL/g?

**Comment by vjpai** [comment]
Nico and I discussed this, and I have a test environment now that uses RH 6.6 and gcc 4.4.x. There is a more official way to do it than what you said - define a class called nullptr_t that supports a templated conversion to any other class or member class. This is better than just using NULL in case we are using it in the context of any functions overloaded on a pointer argument. We can probably introduce this as a grpc::nullptr. Low priority but can be done without too much work.

**Comment by nicolasnoble** [comment]
Just digging a bit, it looks like we might want to try and go a bit deeper than 4.4.x:

http://wiki.centos.org/FAQ/General#head-fe8a0be91ee3e7dea812e8694491e1dde5b75e6d

CentOS5 is supported until 2017.

```
$ sudo docker run -t -i centos:centos5 bash -l
bash-3.2# cat /etc/issue
CentOS release 5.11 (Final)
bash-3.2# yum install gcc
bash-3.2# gcc --version
gcc (GCC) 4.1.2 20080704 (Red Hat 4.1.2-55)
Copyright (C) 2006 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

**Comment by vnitinv** [comment]
@vjpai  I need to first get the grpc setup ready on any OS. Let me get that first. Only after that I can help you with testing/reproducing.

@nicolasnoble 2017 is too far :)

**Comment by vjpai** [comment]
@vnitinv : I was able to dig up an alternative test platform, which is why I picked up this issue now.

**Comment by dklempner** [comment]
@nicolasnoble I'd like to push back at least a little bit, at least regarding centos 5 -- personally, I'd say having C++11isms in the API is far more important than supporting seven year old systems using ten year old compiler branches.

When you run such an old system, you're making a very explicit choice that system stability is more important than being able to run new software, and I suspect that bending over backwards to support such systems isn't a good use of engineering time. gcc 4.4 may be reasonable (and notably includes unique_ptr, which is high on the list of API-relevant C++11isms), but 4.1 definitely isn't.

On the other hand, keep in mind that users of such old platforms have many alternatives, potentially going as far as locally installing a more modern compiler (and potentially libc), or building on another system and statically linking everything.

**Comment by nicolasnoble** [comment]
Oh I'm not saying we should absolutely go at great lengths to support CentOS5 by all costs. I'm just saying it might be worth investigating too.

**Comment by nicolasnoble** [comment]
Also, RedHat's "4.1" gcc isn't really 4.1. It has been patched so much that it has backported features sometime up to 4.5 or 4.6. So it might be worth poking at it just to see what's in there exactly.

**Comment by vnitinv** [comment]
@dklempner David I agree with you but said that it wold be better to specify in the README as which all platform are supported rather then user trying on their device wasting hours and then figuring it out that their device is not supported.

**Comment by nicolasnoble** [comment]
In all fairness, that's not something that can easily be assessed, nor expected from a software released in an alpha state. Getting these feedback and information from early adopters is exactly what we were seeking, and we're taking good notes of what works and what doesn't, in order to properly address these for a non-alpha release.

**Comment by vnitinv** [comment]
@nicolasnoble Thanks Nicolas

**Comment by vjpai** [comment]
I am thinking of considering this infeasible and simply banning gcc versions below 4.6. I have tried 3 different solutions to this (and thought of a fourth), but they all fall apart because we like to use ternary operators and also use a lot of unique_ptr.

1. #define nullptr NULL
This doesn't work because NULL is 0L (a const long int), and there is no implicit conversion of this to unique_ptr . Also all our uses of this in ternary operators with unique_ptr or other pointer types on the other side of the ":" fall apart

2. #define nullptr ((void *)0)
Unlike C, C++ doesn't allow implicit conversion of void * to any arbitrary T *, so there is again an issue with our use of ternary operators and unique_ptr

3. const class {
public:
template <class T> operator T*() const {return static_cast<T *>0;}
operator bool() {return false;}
private:
void operator &() const = delete;
} nullptr = {};
This is suggested as an alternative in the original document proposing nullptr.

unique_ptr is again the problem here. There is no implicit conversion from our anonymous class to unique_ptr, the way that there is from nullptr_t to unique_ptr

4. Make a real class object and have nullptr be a pointer to that.
This will fall apart because conversions to unique_ptr will consider this set, whereas conversions from nullptr to unique_ptr are supposed to be considered unset.

**Comment by vjpai** [comment]
A side note. gcc versions < 4.6 also don't support range-based for loops, which we use in several places in the main code and tests. These are technically just syntactic sugar, but it will again lead to more complicated code (and block compiler optimizations targeted for range-based for)

**Comment by nicolasnoble** [comment]
As for the range-based for loops, they are not supported by Visual Studio
2010 either, so we'll have to get rid of them eventually anyway.

**Comment by vjpai** [comment]
5th attempt

 #define nullptr nulltemplate()
 template <class E> nulltemplate() {return static_cast<E>((void *)0);}

This compiles for the unique_ptr situation as well as most of the ternary operators, but doesn't compile whenever we compare something against nullptr.

**Comment by vjpai** [comment]
I have a working workaround in commit  c89dad3 . There is still one more issue to get the library working with gcc 4.4 - one lambda in thread_pool.cc . That should be easy to fix. There are many lambdas in the tests (particularly qps), so that might be too difficult to fix, but other tests will work fine and the library will be in working shape soon. I intend a pull request about this some time in the next few days.

**Comment by vjpai** [comment]
A variety of problems related to gcc 4.4 has been solved in pull request
number 1135. Please try this out, the main libraries should all work,
though not all of the tests will work.

**Comment by vjpai** [comment]
@vnitinv , can you let us know if you have tried out the code since PR #1135 ?

**Comment by vnitinv** [comment]
@vjpai couldn't got a chance to check this out.I can only do it sometime in next week.

**Comment by vjpai** [comment]
I am closing out this issue for now since I believe that the nullptr issue is now well-addressed. If there's any other issue related to gcc 4.4, let's open it up as a new issue.
